### ISSUE TITLE
MetricsQL `if` doesn't work well

### ISSUE THREAD

**Issue by SunKyu** [VictoriaMetrics/VictoriaMetrics#3993]
### Describe the bug

Hi, I found a bug in VictoriaMetrics v.1.87.3. The Single and Cluster versions are the same. The example queries are as follows:

The query:
`2 if (count(vm_active_merges{job="victoriametrics", type="storage/big"}) == 1)`
returns a series, and
The query:
`2 if (count(vm_active_merges{job="victoriametrics", type="storage/big"}) by (type) == 1)`
returns nothing. The only difference is the grouping part, and the issue is reproduced when I use the 'if' clause. I believe the second query should return the same result as the first query.

And there were no such problem with old version : v1.64.1.
Thank you in advance! 

### To Reproduce

Use the Query from describe.
I just use the VictoriaMetrics metric for easy reproduce

### Version

v.1.87.3. 

### Screenshots
<img width="1537" alt="Screenshot 2023-03-23 at 0 54 29" src="https://user-images.githubusercontent.com/7378727/226973523-709edee7-f8f6-4308-81bb-98c88fd4aa5d.png">

please Check this thread from slack too : https://victoriametrics.slack.com/archives/CGZF1H6L9/p1679482254279959

**Comment by sonjoosik** [comment]
These are the screenshots of traces. 

![Screen Shot 2023-03-23 at 9 37 02 AM](https://user-images.githubusercontent.com/13982648/227069450-8d3ba47e-f6fb-4164-839b-ae946e765391.png)

* The left trace (correct response)
~~~json
{
  "duration_msec": 7.626,
  "message": "victoria-metrics-20230313-063122-tags-v1.87.3-0-g16be82b95: /prometheus/api/v1/query: query=2 if (count(vm_active_merges{job=\"victoriametrics\", type=\"storage/big\"}) == 1), time=1679531636575: series=1",
  "children": [
    {
      "duration_msec": 7.553,
      "message": "eval: query=2 if (count(vm_active_merges{job=\"victoriametrics\", type=\"storage/big\"}) == 1), timeRange=[2023-03-23T00:33:56.575Z..2023-03-23T00:33:56.575Z], step=60000, mayCache=false: series=1, points=1, pointsPerSeries=1",
      "children": [
        {
          "duration_msec": 7.55,
          "message": "binary op \"if\": series=1",
          "children": [
            {
              "duration_msec": 7.501,
              "message": "eval: query=count(vm_active_merges{job=\"victoriametrics\", type=\"storage/big\"}) == 1, timeRange=[2023-03-23T00:33:56.575Z..2023-03-23T00:33:56.575Z], step=60000, mayCache=false: series=1, points=1, pointsPerSeries=1",
              "children": [
                {
                  "duration_msec": 7.5,
                  "message": "binary op \"==\": series=1",
                  "children": [
                    {
                      "duration_msec": 7.496,
                      "message": "execute left and right sides of \"==\" in parallel",
                      "children": [
                        {
                          "duration_msec": 7.486,
                          "message": "expr1",
                          "children": [
                            {
                              "duration_msec": 7.442,
                              "message": "eval: query=count(vm_active_merges{job=\"victoriametrics\", type=\"storage/big\"}), timeRange=[2023-03-23T00:33:56.575Z..2023-03-23T00:33:56.575Z], step=60000, mayCache=false: series=1, points=1, pointsPerSeries=1",
                              "children": [
                                {
                                  "duration_msec": 7.439,
                                  "message": "aggregate count(): series=1",
                                  "children": [
                                    {
                                      "duration_msec": 7.417,
                                      "message": "rollup default_rollup(): timeRange=[2023-03-23T00:33:56.575Z..2023-03-23T00:33:56.575Z], step=60000, window=0: neededMemoryBytes=1016",
                                      "children": [
                                        {
                                          "duration_msec": 0.001,
                                          "message": "rollup cache get: query=count(vm_active_merges{job=\"victoriametrics\", type=\"storage/big\"}), timeRange=[2023-03-23T00:33:56.575Z..2023-03-23T00:33:56.575Z], step=60000, window=0",
                                          "children": [
                                            {
                                              "duration_msec": 0,
                                              "message": "do not fetch series from cache, since it is disabled in the current context"
                                            }
                                          ]
                                        },
                                        {
                                          "duration_msec": 7.252,
                                          "message": "fetch matching series: filters=[{__name__=\"vm_active_merges\",job=\"victoriametrics\",type=\"storage/big\"}], timeRange=[2023-03-23T00:27:56.575Z..2023-03-23T00:33:56.575Z]",
                                          "children": [
                                            {
                                              "duration_msec": 6.587,
                                              "message": "init series search: filters=[{__name__=\"vm_active_merges\",job=\"victoriametrics\",type=\"storage/big\"}], timeRange=[2023-03-23T00:27:56.575Z..2023-03-23T00:33:56.575Z]",
                                              "children": [
                                                {
                                                  "duration_msec": 6.07,
                                                  "message": "search for matching metricIDs: filters=[{__name__=\"vm_active_merges\",job=\"victoriametrics\",type=\"storage/big\"}], timeRange=[2023-03-23T00:27:56.575Z..2023-03-23T00:33:56.575Z]",
                                                  "children": [
                                                    {
                                                      "duration_msec": 6.042,
                                                      "message": "search for metricIDs in the current indexdb",
                                                      "children": [
                                                        {
                                                          "duration_msec": 0.007,
                                                          "message": "search for metricIDs in tag filters cache",
                                                          "children": [
                                                            {
                                                              "duration_msec": 0.006,
                                                              "message": "cache miss"
                                                            }
                                                          ]
                                                        },
                                                        {
                                                          "duration_msec": 3.896,
                                                          "message": "search for metric ids: filters=[{composite(vm_active_merges,job)=\"victoriametrics\",composite(vm_active_merges,type)=\"storage/big\"}], timeRange=[2023-03-23T00:27:56.575Z..2023-03-23T00:33:56.575Z], maxMetrics=3000000",
                                                          "children": [
                                                            {
                                                              "duration_msec": 3.89,
                                                              "message": "update metric ids: filters={composite(vm_active_merges,job)=\"victoriametrics\",composite(vm_active_merges,type)=\"storage/big\"}, timeRange=[2023-03-23T00:27:56.575Z..2023-03-23T00:33:56.575Z]: updated 1 metric ids",
                                                              "children": [
                                                                {
                                                                  "duration_msec": 3.884,
                                                                  "message": "search for metric ids on a particular day: filters={composite(vm_active_merges,job)=\"victoriametrics\",composite(vm_active_merges,type)=\"storage/big\"}, date=2023-03-23, maxMetrics=3000001",
                                                                  "children": [
                                                                    {
                                                                      "duration_msec": 3.686,
                                                                      "message": "search for the first non-negative filter with the smallest cost",
                                                                      "children": [
                                                                        {
                                                                          "duration_msec": 3.678,
                                                                          "message": "get metric ids for filter and date: filter={composite(vm_active_merges,type)=\"storage/big\"}, date=2023-03-23, maxMetrics=150000050, maxLoopsCount=5",
                                                                          "children": [
                                                                            {
                                                                              "duration_msec": 3.67,
                                                                              "message": "found 1 metric ids for filter={composite(vm_active_merges,type)=\"storage/big\"} using exact search; spent 1 loops"
                                                                            }
                                                                          ]
                                                                        },
                                                                        {
                                                                          "duration_msec": 0,
                                                                          "message": "the filter={composite(vm_active_merges,type)=\"storage/big\"} matches less than 150000050 series (actually 1 series); use it"
                                                                        }
                                                                      ]
                                                                    },
                                                                    {
                                                                      "duration_msec": 0.187,
                                                                      "message": "intersect the remaining 1 filters with the found 1 metric ids",
                                                                      "children": [
                                                                        {
                                                                          "duration_msec": 0.031,
                                                                          "message": "get metric ids for filter and date: filter={composite(vm_active_merges,job)=\"victoriametrics\"}, date=2023-03-23, maxMetrics=9223372036854775807, maxLoopsCount=150",
                                                                          "children": [
                                                                            {
                                                                              "duration_msec": 0.029,
                                                                              "message": "found 5 metric ids for filter={composite(vm_active_merges,job)=\"victoriametrics\"} using exact search; spent 5 loops"
                                                                            }
                                                                          ]
                                                                        },
                                                                        {
                                                                          "duration_msec": 0.148,
                                                                          "message": "intersect 5 metric ids with the found 1 metric ids for filter={composite(vm_active_merges,job)=\"victoriametrics\"}; resulting metric ids: 1"
                                                                        }
                                                                      ]
                                                                    },
                                                                    {
                                                                      "duration_msec": 0,
                                                                      "message": "found 1 metric ids"
                                                                    }
                                                                  ]
                                                                }
                                                              ]
                                                            }
                                                          ]
                                                        },
                                                        {
                                                          "duration_msec": 0,
                                                          "message": "sort 1 matching metric ids"
                                                        }
                                                      ]
                                                    },
                                                    {
                                                      "duration_msec": 0.008,
                                                      "message": "search for metricIDs in the previous indexdb",
                                                      "children": [
                                                        {
                                                          "duration_msec": 0.004,
                                                          "message": "search for metricIDs in tag filters cache",
                                                          "children": [
                                                            {
                                                              "duration_msec": 0.003,
                                                              "message": "found metricIDs with size: 8 bytes"
                                                            },
                                                            {
                                                              "duration_msec": 0,
                                                              "message": "unmarshaled 0 metricIDs"
                                                            }
                                                          ]
                                                        }
                                                      ]
                                                    },
                                                    {
                                                      "duration_msec": 0,
                                                      "message": "merge 1 metricIDs from the current indexdb with 0 metricIDs from the previous indexdb; result: 1 metricIDs"
                                                    },
                                                    {
                                                      "duration_msec": 0.01,
                                                      "message": "put 1 metricIDs in cache",
                                                      "children": [
                                                        {
                                                          "duration_msec": 0,
                                                          "message": "marshaled 1 metricIDs into 16 bytes"
                                                        },
                                                        {
                                                          "duration_msec": 0.009,
                                                          "message": "stored 1 metricIDs into cache"
                                                        }
                                                      ]
                                                    }
                                                  ]
                                                },
                                                {
                                                  "duration_msec": 0.01,
                                                  "message": "obtain tsids from 1 metricIDs",
                                                  "children": [
                                                    {
                                                      "duration_msec": 0.003,
                                                      "message": "found 1 tsids for 1 metricIDs in the current indexdb"
                                                    },
                                                    {
                                                      "duration_msec": 0.004,
                                                      "message": "found 0 tsids for 0 metricIDs in the previous indexdb"
                                                    },
                                                    {
                                                      "duration_msec": 0,
                                                      "message": "load 1 tsids for 1 metricIDs from both current and previous indexdb"
                                                    },
                                                    {
                                                      "duration_msec": 0.001,
                                                      "message": "sort 1 tsids"
                                                    }
                                                  ]
                                                },
                                                {
                                                  "duration_msec": 0.002,
                                                  "message": "prefetch metric names for 1 metricIDs",
                                                  "children": [
                                                    {
                                                      "duration_msec": 0,
                                                      "message": "1 out of 1 metric names must be pre-fetched"
                                                    },
                                                    {
                                                      "duration_msec": 0,
                                                      "message": "skip pre-fetching metric names for low number of metrid ids=1"
                                                    }
                                                  ]
                                                },
                                                {
                                                  "duration_msec": 0.495,
                                                  "message": "search for parts with data for 1 series"
                                                }
                                              ]
                                            },
                                            {
                                              "duration_msec": 0.619,
                                              "message": "fetch unique series=1, blocks=9, samples=126, bytes=729"
                                            }
                                          ]
                                        },
                                        {
                                          "duration_msec": 0.12,
                                          "message": "rollup default_rollup() with incremental aggregation count() over 1 series; rollupConfigs=[timeRange=[2023-03-23T00:33:56.575Z..2023-03-23T00:33:56.575Z], step=60000, window=0, points=1]",
                                          "children": [
                                            {
                                              "duration_msec": 0.109,
                                              "message": "parallel process of fetched data: series=1, samples=36"
                                            },
                                            {
                                              "duration_msec": 0.006,
                                              "message": "series after aggregation with count(): 1; samplesScanned=37"
                                            }
                                          ]
                                        },
                                        {
                                          "duration_msec": 0.001,
                                          "message": "rollup cache put: query=count(vm_active_merges{job=\"victoriametrics\", type=\"storage/big\"}), timeRange=[2023-03-23T00:33:56.575Z..2023-03-23T00:33:56.575Z], step=60000, window=0, series=1",
                                          "children": [
                                            {
                                              "duration_msec": 0,
                                              "message": "do not store series to cache, since it is disabled in the current context"
                                            }
                                          ]
                                        }
                                      ]
                                    }
                                  ]
                                }
                              ]
                            }
                          ]
                        },
                        {
                          "duration_msec": 0.011,
                          "message": "expr2",
                          "children": [
                            {
                              "duration_msec": 0.003,
                              "message": "eval: query=1, timeRange=[2023-03-23T00:33:56.575Z..2023-03-23T00:33:56.575Z], step=60000, mayCache=false: series=1, points=1, pointsPerSeries=1"
                            }
                          ]
                        }
                      ]
                    }
                  ]
                }
              ]
            },
            {
              "duration_msec": 0.001,
              "message": "eval: query=2, timeRange=[2023-03-23T00:33:56.575Z..2023-03-23T00:33:56.575Z], step=60000, mayCache=false: series=1, points=1, pointsPerSeries=1"
            }
          ]
        }
      ]
    },
    {
      "duration_msec": 0.001,
      "message": "leave only the first point in every series"
    },
    {
      "duration_msec": 0.001,
      "message": "sort series by metric name and labels"
    },
    {
      "duration_msec": 0.035,
      "message": "generate /api/v1/query response for series=1"
    }
  ]
}
~~~

* The right trace (wrong response)
~~~json
{
  "duration_msec": 0.428,
  "message": "victoria-metrics-20230313-063122-tags-v1.87.3-0-g16be82b95: /prometheus/api/v1/query: query=2 if (count(vm_active_merges{job=\"victoriametrics\", type=\"storage/big\"}) by (type) == 1), time=1679531613816: series=0",
  "children": [
    {
      "duration_msec": 0.394,
      "message": "eval: query=2 if (count(vm_active_merges{job=\"victoriametrics\", type=\"storage/big\"}) by (type) == 1), timeRange=[2023-03-23T00:33:33.816Z..2023-03-23T00:33:33.816Z], step=60000, mayCache=false: series=0, points=0, pointsPerSeries=0",
      "children": [
        {
          "duration_msec": 0.39,
          "message": "binary op \"if\": series=0",
          "children": [
            {
              "duration_msec": 0.373,
              "message": "eval: query=count(vm_active_merges{job=\"victoriametrics\", type=\"storage/big\"}) by (type) == 1, timeRange=[2023-03-23T00:33:33.816Z..2023-03-23T00:33:33.816Z], step=60000, mayCache=false: series=1, points=1, pointsPerSeries=1",
              "children": [
                {
                  "duration_msec": 0.372,
                  "message": "binary op \"==\": series=1",
                  "children": [
                    {
                      "duration_msec": 0.35,
                      "message": "eval: query=count(vm_active_merges{job=\"victoriametrics\", type=\"storage/big\"}) by (type), timeRange=[2023-03-23T00:33:33.816Z..2023-03-23T00:33:33.816Z], step=60000, mayCache=false: series=1, points=1, pointsPerSeries=1",
                      "children": [
                        {
                          "duration_msec": 0.348,
                          "message": "aggregate count(): series=1",
                          "children": [
                            {
                              "duration_msec": 0.34,
                              "message": "rollup default_rollup(): timeRange=[2023-03-23T00:33:33.816Z..2023-03-23T00:33:33.816Z], step=60000, window=0: neededMemoryBytes=1016",
                              "children": [
                                {
                                  "duration_msec": 0.001,
                                  "message": "rollup cache get: query=count(vm_active_merges{job=\"victoriametrics\", type=\"storage/big\"}) by (type), timeRange=[2023-03-23T00:33:33.816Z..2023-03-23T00:33:33.816Z], step=60000, window=0",
                                  "children": [
                                    {
                                      "duration_msec": 0,
                                      "message": "do not fetch series from cache, since it is disabled in the current context"
                                    }
                                  ]
                                },
                                {
                                  "duration_msec": 0.256,
                                  "message": "fetch matching series: filters=[{__name__=\"vm_active_merges\",job=\"victoriametrics\",type=\"storage/big\"}], timeRange=[2023-03-23T00:27:33.816Z..2023-03-23T00:33:33.816Z]",
                                  "children": [
                                    {
                                      "duration_msec": 0.229,
                                      "message": "init series search: filters=[{__name__=\"vm_active_merges\",job=\"victoriametrics\",type=\"storage/big\"}], timeRange=[2023-03-23T00:27:33.816Z..2023-03-23T00:33:33.816Z]",
                                      "children": [
                                        {
                                          "duration_msec": 0.013,
                                          "message": "search for matching metricIDs: filters=[{__name__=\"vm_active_merges\",job=\"victoriametrics\",type=\"storage/big\"}], timeRange=[2023-03-23T00:27:33.816Z..2023-03-23T00:33:33.816Z]",
                                          "children": [
                                            {
                                              "duration_msec": 0.008,
                                              "message": "search for metricIDs in the current indexdb",
                                              "children": [
                                                {
                                                  "duration_msec": 0.006,
                                                  "message": "search for metricIDs in tag filters cache",
                                                  "children": [
                                                    {
                                                      "duration_msec": 0.004,
                                                      "message": "found metricIDs with size: 16 bytes"
                                                    },
                                                    {
                                                      "duration_msec": 0.001,
                                                      "message": "unmarshaled 1 metricIDs"
                                                    }
                                                  ]
                                                }
                                              ]
                                            }
                                          ]
                                        },
                                        {
                                          "duration_msec": 0.01,
                                          "message": "obtain tsids from 1 metricIDs",
                                          "children": [
                                            {
                                              "duration_msec": 0.005,
                                              "message": "found 1 tsids for 1 metricIDs in the current indexdb"
                                            },
                                            {
                                              "duration_msec": 0.001,
                                              "message": "found 0 tsids for 0 metricIDs in the previous indexdb"
                                            },
                                            {
                                              "duration_msec": 0,
                                              "message": "load 1 tsids for 1 metricIDs from both current and previous indexdb"
                                            },
                                            {
                                              "duration_msec": 0.001,
                                              "message": "sort 1 tsids"
                                            }
                                          ]
                                        },
                                        {
                                          "duration_msec": 0.002,
                                          "message": "prefetch metric names for 1 metricIDs",
                                          "children": [
                                            {
                                              "duration_msec": 0,
                                              "message": "1 out of 1 metric names must be pre-fetched"
                                            },
                                            {
                                              "duration_msec": 0,
                                              "message": "skip pre-fetching metric names for low number of metrid ids=1"
                                            }
                                          ]
                                        },
                                        {
                                          "duration_msec": 0.195,
                                          "message": "search for parts with data for 1 series"
                                        }
                                      ]
                                    },
                                    {
                                      "duration_msec": 0.01,
                                      "message": "fetch unique series=1, blocks=11, samples=123, bytes=891"
                                    }
                                  ]
                                },
                                {
                                  "duration_msec": 0.049,
                                  "message": "rollup default_rollup() with incremental aggregation count() over 1 series; rollupConfigs=[timeRange=[2023-03-23T00:33:33.816Z..2023-03-23T00:33:33.816Z], step=60000, window=0, points=1]",
                                  "children": [
                                    {
                                      "duration_msec": 0.04,
                                      "message": "parallel process of fetched data: series=1, samples=36"
                                    },
                                    {
                                      "duration_msec": 0.005,
                                      "message": "series after aggregation with count(): 1; samplesScanned=37"
                                    }
                                  ]
                                },
                                {
                                  "duration_msec": 0.001,
                                  "message": "rollup cache put: query=count(vm_active_merges{job=\"victoriametrics\", type=\"storage/big\"}) by (type), timeRange=[2023-03-23T00:33:33.816Z..2023-03-23T00:33:33.816Z], step=60000, window=0, series=1",
                                  "children": [
                                    {
                                      "duration_msec": 0,
                                      "message": "do not store series to cache, since it is disabled in the current context"
                                    }
                                  ]
                                }
                              ]
                            }
                          ]
                        }
                      ]
                    },
                    {
                      "duration_msec": 0.001,
                      "message": "eval: query=1, timeRange=[2023-03-23T00:33:33.816Z..2023-03-23T00:33:33.816Z], step=60000, mayCache=false: series=1, points=1, pointsPerSeries=1"
                    }
                  ]
                }
              ]
            },
            {
              "duration_msec": 0.001,
              "message": "eval: query=2, timeRange=[2023-03-23T00:33:33.816Z..2023-03-23T00:33:33.816Z], step=60000, mayCache=false: series=1, points=1, pointsPerSeries=1"
            }
          ]
        }
      ]
    },
    {
      "duration_msec": 0.001,
      "message": "leave only the first point in every series"
    },
    {
      "duration_msec": 0.001,
      "message": "sort series by metric name and labels"
    },
    {
      "duration_msec": 0.004,
      "message": "generate /api/v1/query response for series=0"
    }
  ]
}
~~~

**Comment by dmitryk-dk** [comment]
Hi @SunKyu, @sonjoosik  ! The problem was introduced by this (commit)[https://github.com/VictoriaMetrics/VictoriaMetrics/commit/b5da47bfaf2f573853645030bcf1ff0db331d31a]. I will try to investigate and fix soon

**Comment by valyala** [comment]
The `a if b` query returns `a` results only if `b` contains non-empty time series with the identical set of labels as at the `a` side.

The query `2 if (count(vm_active_merges{job="victoriametrics", type="storage/big"}) == 1)` returns non-empty results because `count(vm_active_merges{job="victoriametrics", type="storage/big"}) == 1` returns a single time series without any labels. The `2` query also contains zero labels. So it matches the right-hand side in the `if` query.

The query `2 if (count(vm_active_merges{job="victoriametrics", type="storage/big"}) by (type) == 1)` returns nothing, since `count(vm_active_merges{job="victoriametrics", type="storage/big"}) by (type) == 1` returns a time series with `type` label. This doesn't match zero labels for `2` query. The solution is to add `on()` modifier to the query, so it uses an empty set of labels for searching for the matching time series on the right side of `if` query:

```
2 if on() (count(vm_active_merges{job="victoriametrics", type="storage/big"}) by (type) == 1)
```

See [docs about on() modifier](https://prometheus.io/docs/prometheus/latest/querying/operators/#vector-matching-keywords).

BTW, the query `count(vm_active_merges{job="victoriametrics", type="storage/big"})` looks unusual - it always returns the constant number of matching time series. If you want calculating the number of active merges across multiple VictoriaMetrics instances (or vmstorage instances in [VictoriaMetrics cluster](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html)), then `count()` must be substituted with `sum()`:

```
sum(vm_active_merges{job="victoriametrics", type="storage/big"})
```

See [count()](https://docs.victoriametrics.com/MetricsQL.html#count) and [sum()](https://docs.victoriametrics.com/MetricsQL.html#sum) docs for details.

**Comment by sonjoosik** [comment]
@valyala 
Thank you for the comprehensive explanation! I now understand why the difference occurred, and I believe that the solution using the on() modifier can address our issue. I appreciate your help!

Regarding the unusual query, it was created solely for reproducing the problem we encountered. We had an issue with another, rather lengthy query that included an if clause. So, please don't be concerned about the example query. Thank you for your thoughtful guidance.

**Comment by dmitryk-dk** [comment]
Close as resolved. Another issue was created
